src/Nethermind/Nethermind.Trie/TrieNode.cs

```diff
@@ -895,29 +905,36 @@ public void CallRecursively(
                 return;
             }
 
-            if (!IsLeaf)
+            if (_nodeData is BranchData branchData)
             {
-                object?[] data = _data;
-                if (data is not null)
+                ref readonly var data = ref branchData.Branches;
+                for (int i = 0; i < BranchArray.Length; i++)
                 {
-                    for (int i = 0; i < data.Length; i++)
+                    if (data[i] is TrieNode child)
                     {
-                        if (data[i] is TrieNode child)
-                        {
-                            if (logger.IsTrace) logger.Trace($"Persist recursively on child {i} {child} of {this}");
-                            int previousLength = AppendChildPath(ref currentPath, i);
-                            child.CallRecursively(action, storageAddress, ref currentPath, resolver, skipPersisted, logger, maxPathLength, resolveStorageRoot);
-                            currentPath.TruncateMut(previousLength);
-                        }
+                        if (logger.IsTrace) logger.Trace($"Persist recursively on child {i} {child} of {this}");
+                        int previousLength = AppendChildPath(ref currentPath, i);
```

Comment:
NP: You can append once and just change the path since you know its a branch. See:
https://github.com/NethermindEth/nethermind/pull/7915/files#diff-38006f905d7c66102843b20881e250f1465a7be3fd18cee6f0a3e28ef08f9ca9L1254